code/game/objects/items/rcd/RPD.dm

```diff
@@ -273,15 +298,114 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
+/obj/item/pipe_dispenser/AltClick(mob/user)
+	. = ..()
+	if(!(upgrade_flags & RPD_UPGRADE_SPEEDPIPE))
+		return .
+	toggle_speedpipe_marking(user)
+
+/obj/item/pipe_dispenser/alt_click_secondary(mob/user)
+	. = ..()
+	speedpipe_build(user)
+
+/obj/item/pipe_dispenser/CtrlClick(mob/user)
+	. = ..()
+	clear_out_speedpipe(user)
+
+/obj/item/pipe_dispenser/proc/toggle_speedpipe_marking(mob/user)
+	speedpipe_marking = !speedpipe_marking
+	balloon_alert(user, "[speedpipe_marking ? "marking speedpipe locations" : "stopped marking"]")
+
+/obj/item/pipe_dispenser/proc/speedpipe_build(mob/user)
+	if(length(speedpipe_turfs))
+		speedpipe_building = !speedpipe_building
+		balloon_alert(user, "[speedpipe_building ? "speedpiping..." : "stopped speedpiping"]")
+		if(!speedpipe_building)
+			return
+		deltimer(timed_out_speedpipe_timer)
+		clear_orphaned_speedpipe_turfs()
+		for(var/turf/speedpipe_turf as anything in speedpipe_turfs)
+			if(!speedpipe_building)
+				return
+			if(speedpipe_turf.type == old_turfs[speedpipe_turf])
+				do_rpd_actions(speedpipe_turf, user, speedpipe_build = TRUE)
+				speedpipe_turf.cut_overlay(rpd_overlay)
+				speedpipe_turfs -= speedpipe_turf
+				old_turfs -= speedpipe_turf
+
+/obj/item/pipe_dispenser/pickup(mob/to_hook)
+	. = ..()
+	if(!(upgrade_flags & RPD_UPGRADE_SPEEDPIPE))
+		return .
+	if(listeningTo == to_hook)
+		return .
+	if(listeningTo)
+		UnregisterSignal(listeningTo, COMSIG_MOVABLE_MOVED)
+	RegisterSignal(to_hook, COMSIG_MOVABLE_MOVED, PROC_REF(on_move))
+	listeningTo = to_hook
+
+/obj/item/pipe_dispenser/proc/on_move(mob/user)
+	SIGNAL_HANDLER
+	if(!speedpipe_marking)
+		return
+	if(!isturf(user.loc))
+		return
+	var/turf/the_turf = user.loc
+	if(!(the_turf in speedpipe_turfs))
+		speedpipe_turfs += the_turf
+		old_turfs += the_turf
+		old_turfs[the_turf] = the_turf.type
+		var/old_turf_color = the_turf.color
+		the_turf.color = COLOR_DARK_CYAN
+		animate(the_turf, 2 SECONDS, color = old_turf_color)
+		the_turf.add_overlay(rpd_overlay)
+		addtimer(CALLBACK(the_turf, TYPE_PROC_REF(/atom, update_atom_colour)), 2 SECONDS)
+		deltimer(timed_out_speedpipe_timer)
+		timed_out_speedpipe_timer = addtimer(CALLBACK(src, PROC_REF(clear_out_speedpipe), user), 3 MINUTES, TIMER_STOPPABLE)
+
+/obj/item/pipe_dispenser/proc/clear_orphaned_speedpipe_turfs()
+	for(var/turf/speedpipe_turf as anything in speedpipe_turfs)
+		if(QDELETED(speedpipe_turf))
+			speedpipe_turfs -= speedpipe_turf
+	for(var/turf/speedpipe_turf as anything in old_turfs)
+		if(QDELETED(speedpipe_turf))
+			old_turfs -= speedpipe_turf
+
+/obj/item/pipe_dispenser/proc/clear_out_speedpipe(mob/user, list/the_speedpipe_turfs = speedpipe_turfs)
+	balloon_alert(user, "cleared speedpipe markings")
+	for(var/turf/speedpipe_turf as anything in the_speedpipe_turfs)
+		speedpipe_turf.cut_overlay(rpd_overlay)
+		speedpipe_turfs -= speedpipe_turf
+		old_turfs -= speedpipe_turf
+	the_speedpipe_turfs = list()
+	if(the_speedpipe_turfs == speedpipe_turfs)
```

Comment:
Not only do I have no idea what the_speedpipe_turfs = list() is supposed to do here, but also this if check will never ever pass because the `= list()` will always create a new list, which will fail equality check.

Did you mean `.Cut()`?